logic/src/androidMain/kotlin/com/wire/kalium/logic/network/NetworkStateObserverImpl.kt

```diff
@@ -0,0 +1,90 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.network
+
+import android.app.Activity
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.shareIn
+
+actual class NetworkStateObserverImpl(
+    appContext: Context,
+    kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl
+) : NetworkStateObserver {
+    private val connectivityManager: ConnectivityManager = appContext.getSystemService(Activity.CONNECTIVITY_SERVICE) as ConnectivityManager
+    private val scope = CoroutineScope(SupervisorJob() + kaliumDispatcher.io)
+    private val networkStateSharedFlow: Flow<NetworkState> = callbackFlow {
+        val callback = object : ConnectivityManager.NetworkCallback() {
+
+            override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
+                super.onCapabilitiesChanged(network, networkCapabilities)
+                trySend(networkCapabilities.toState())
+            }
+
+            override fun onBlockedStatusChanged(network: Network, blocked: Boolean) {
+                super.onBlockedStatusChanged(network, blocked)
+                trySend(if (blocked) NetworkState.ConnectedWithoutInternet else NetworkState.ConnectedWithInternet)
+            }
+
+            override fun onLost(network: Network) {
+                trySend(NetworkState.NotConnected)
+                super.onLost(network)
+            }
+
+            override fun onUnavailable() {
+                trySend(NetworkState.NotConnected)
+                super.onUnavailable()
+            }
+        }
+
+        trySend(connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork).toState())
+
+        connectivityManager.registerDefaultNetworkCallback(callback)
+
+        awaitClose {
+            connectivityManager.unregisterNetworkCallback(callback)
+        }
+    }.shareIn(
+        scope = scope,
+        started = SharingStarted.WhileSubscribed(),
+        replay = 1
+    )
+
+    private fun NetworkCapabilities?.toState(): NetworkState {
+        val hasInternet = this?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
+        val isValidated = this?.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED) == true
```

Comment:
I wonder if there are some edge cases where on-premise environments could be considered "not validated" and should still be able to make requests.

Based on [the docs](https://developer.android.com/training/basics/network-ops/reading-network-state#introducing-net-capabilities):

> [NET_CAPABILITY_VALIDATED](https://developer.android.com/reference/android/net/NetworkCapabilities#NET_CAPABILITY_VALIDATED): This capability indicates the network has been found to provide actual access to the public internet last time it was probed.

Some customers MIGHT be affected, but I believe this is a great first iteration anyway and this can continue the way it is for now. We can add a `TODO` or just regular comment for now and check with customers.

Reply:
Ok, I added a comment to keep that in mind 😉